src/erl.rs

```diff
@@ -1492,52 +1517,52 @@ impl TypeAst {
-
+            TypeAst::Tuple { elems, .. } => tuple(elems.iter().map(|e| e.to_erlang_type_def())),
             TypeAst::Constructor {
                 module, name, args, ..
             } => {
-                let args = if args.is_empty() {
-                    nil()
-                } else {
-                    concat(
-                        args.iter()
-                            .map(|a| a.to_erlang_type_def())
-                            .intersperse(", ".to_doc()),
-                    )
-                };
+                let arg_list = args
+                    .iter()
+                    .map(|a| a.to_erlang_type_def())
+                    .intersperse(", ".to_doc());
                 if let Some(mod_name) = &module {
-                    let t = mod_name
+                    mod_name
                         .clone()
                         .to_doc()
                         .append(".")
-                        .append(atom(name.clone()));
-                    if args.is_nil() {
-                        t.append("()")
-                    } else {
-                        t.append("(").append(args).append(")")
-                    }
+                        .append(atom(name.clone()))
+                        .append("(")
+                        .append(concat(arg_list))
+                        .append(")")
                 } else {
                     match name.as_str() {
-                        // TODO: more builtins ? Result?
+                        // TODO: more builtins ?
                         "Int" => "integer()".to_doc(),
-                        "String" => "string()".to_doc(),
+                        "String" => "unicode::unicode_binary()".to_doc(),
                         "Bool" => "boolean()".to_doc(),
                         "Float" => "float()".to_doc(),
-                        "List" => "list(".to_doc().append(args).append(")"),
-                        _ => variable_name(name).to_doc(),
+                        "List" => "list(".to_doc().append(concat(arg_list)).append(")"),
+                        "BitString" => "bitstring()".to_doc(),
+                        "Result" => {
+                            let arg_ok = args[0].to_erlang_type_def();
+                            let arg_err = args[1].to_erlang_type_def();
+                            concat(
+                                vec![
+                                    tuple(vec!["ok".to_doc(), arg_ok].into_iter()),
+                                    tuple(vec!["error".to_doc(), arg_err].into_iter()),
+                                ]
+                                .into_iter()
+                                .intersperse(break_(" |", " | ")),
+                            )
+                            .nest(INDENT)
+                            .group()
+                        }
+                        // QUESTION: what should we do here?
+                        // I think after builtins this would be external types?
```

Comment:
I think this should be unreachable. Let's call `fatal_compiler_bug` as something has gone very wrong and we would like to know about it.

Reply:
This code is actually reached when an external function returns an external type. This is part of bullet 3 in my post above. Looking deeper 🤿.

Reply:
This is a bug. I'll try and find some time to investigate tomorrow.